### PR TITLE
APIGOV-20810 - add interface methods for request data

### PR DESCRIPTION
**Pull request by jcollins-axway**
- Access Request data, which matches the schema in the access request definition
- Credential data, which matches the request schema in the credential request definition
- Public key oauth support

### CODE REVIEW

**Review by sbolosan — APPROVED** [review]

**Review by vivekschauhan — APPROVED** [review]

File: pkg/apic/provisioning/credentialrequest.go [hunk 1]
```diff
@@ -10,4 +10,6 @@ type CredentialRequest interface {
 	GetCredentialDetailsValue(key string) string
 	// GetCredentialType returns the type of credential related to this request
 	GetCredentialType() string
+	// GetCredentialData returns the map[string]interface{} of data from the request
+	GetCredentialData() map[string]interface{}
```
Comment (jcollins-axway) [inline pkg/apic/provisioning/credentialrequest.go:14]: @tsjohns9 this change ant he GetAccessRequestData above are meant to return the data that matches the request schema in the *RequestDefinition resources

File: pkg/agent/provisioning.go [hunk 2]
```diff
@@ -34,46 +28,118 @@ func createOrUpdateCredentialRequestDefinition(data *v1alpha1.CredentialRequestD
...
+type crdBuilderOptions struct {
+	name      string
+	provProps []provisioning.PropertyBuilder
+	reqProps  []provisioning.PropertyBuilder
+}
+
 // NewCredentialRequestBuilder - called by the agents to build and register a new credential reqest definition
-func NewCredentialRequestBuilder() provisioning.CredentialRequestBuilder {
-	return provisioning.NewCRDBuilder(createOrUpdateCredentialRequestDefinition)
+func NewCredentialRequestBuilder(options ...func(*crdBuilderOptions)) provisioning.CredentialRequestBuilder {
```
Comment (tsjohns9) [inline pkg/agent/provisioning.go:38]: A specific type might be helpful for these funcs

File: pkg/agent/provisioning.go [hunk 3]
```diff
@@ -34,46 +28,118 @@ func createOrUpdateCredentialRequestDefinition(data *v1alpha1.CredentialRequestD
...
-		SetName(apikeyCRD).
-		SetProvisionSchema(provisioning.NewSchemaBuilder().
-			AddProperty(
-				provisioning.NewSchemaPropertyBuilder().
-					SetName("key").
-					SetLabel("API Key").
-					SetRequired().
-					IsString().
-					IsEncrypted()))
+func NewAPIKeyCredentialRequestBuilder(options ...func(*crdBuilderOptions)) provisioning.CredentialRequestBuilder {
```
Comment (tsjohns9) [inline pkg/agent/provisioning.go:112]: Do agents need to be updated?

**Review by tsjohns9 — APPROVED** [review]

File: pkg/apic/provisioning/definitions.go [hunk 4]
```diff
@@ -1,5 +1,19 @@
...
+	APIKeyARD         = "api-key"
+	APIKeyCRD         = "api-key"
+	OAuthSecretCRD    = "oauth-secret"
+	OAuthPublicKeyCRD = "oauth-public-key"
+
+	OauthClientID     = "clientId"
+	OauthClientSecret = "clientSecret"
+	OauthPublicKey    = "publicKey"
+
+	APIKey = "apiKey"
```
[the remark this answers is not in the document]
Comment (jcollins-axway) [reply pkg/apic/provisioning/definitions.go:14]: The "api-key" here is the name of the resources, as we tend to name resources separated by dashes. The "apiKey" is the field name in the request definition as we tend to name fields in camel case.

File: pkg/agent/provisioning.go (hunk 3 above)
Comment (jcollins-axway) [reply pkg/agent/provisioning.go:112]: Unless you need to add options, not for this call.

File: pkg/agent/provisioning.go (hunk 2 above)
Comment (tsjohns9) [reply pkg/agent/provisioning.go:38]: Actually, I'm not sure how an agent can implement this func because the param is a private type

Comment (jcollins-axway) [reply pkg/agent/provisioning.go:38]: They use the With functions that return that function. They do not need to know that type.

Comment (tsjohns9) [reply pkg/agent/provisioning.go:38]: I see a few exported funcs like `WithCRDProvisionSchemaProperty ` that return `func(c *crdBuilderOptions)` that an agent can use. But if an agent needs their own option func then they won't be able to implement it. Not sure if they need it or not, just bringing it up.

Comment (jcollins-axway) [reply pkg/agent/provisioning.go:38]: Nah, I think we will be the only ones providing these options

**Review by tsjohns9 — APPROVED** [review]

**Pull request merged**